### ISSUE TITLE
Posts duplicated and out of order when viewed in Ivory for iOS client

### ISSUE THREAD

**Issue by dvydra** [jointakahe/takahe#505]
Running takahe 0.7 as per the standard installation instructions, over the past twelve hours I have noticed that Ivory for iOS is seeing many repeated copies of posts in my timeline, and the posts retrieved overnight do not seem to be in chronological order. 

The problem does not occur on the website view of the timeline. 

Happy to share logs, etc as will be helpful.

**Comment by dvydra** [comment]
I’ve also raised this with Ivory https://social.ivydra.com/@daniel/posts/147352384671982232/

**Comment by andrewgodwin** [comment]
Yeah, this is a known problem with Ivory specifically where something they're doing with pagination is exposing a slight incompatability on our end, but I don't know what it is yet (and it only seems to happen after a decent time gap).

**Comment by dvydra** [comment]
Thanks, sorry I looked for client specific stuff before creating this issue, but couldn't find anything. If there's any debugging I can do on the takahe side please let me know

**Comment by andrewgodwin** [comment]
It just needs someone to sit down and try fiddling with the pagination code with Ivory connected to a dev server. I have all I need to do that, I just need to find the time to do it (it will not be for the next week and a bit while I am on holiday!)

**Comment by dvydra** [comment]
I'm happy to debug this, is there a good way to view inbound API requests to my instance? I have multiple clients and some of them are ok, and other exhibit the same issues as Ivory.

**Comment by dvydra** [comment]
Some client interactions against takahe 0.8: 

These were all captured with with an already up to date timeline. I will pull these again in the morning to see which ones get weird with a lot of timeline to catch up on.


# ivory 1.2
## on startup
"GET /api/v1/timelines/home?limit=40&min_id=152252244464626184 HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/preferences HTTP/1.1" 404 3494 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v2/instance HTTP/1.1" 200 2147 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/lists HTTP/1.1" 404 3494 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/notifications?limit=80&min_id=2340&types%5B%5D=mention HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/notifications?limit=80&min_id=2340&types%5B%5D=mention&types%5B%5D=status&types%5B%5D=reblog&types%5B%5D=follow&types%5B%5D=follow_request&types%5B%5D=favourite&types%5B%5D=poll&types%5B%5D=update HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/accounts/147050034222768154/statuses?exclude_reblogs=1&limit=40 HTTP/1.1" 200 17668 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/accounts/147050034222768154/statuses?exclude_reblogs=1&limit=40&max_id=147051308662802792 HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"POST /inbox/ HTTP/1.1" 202 0 "-" "http.rb/5.1.1 (Mastodon/4.1.0; +https://mstdn.social/)"
"GET /api/v1/streaming/public/local HTTP/1.1" 404 3578 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:106.0) Gecko/20100101 Firefox/106.0"


## about once a minute on while idle on home timeline
"GET /api/v1/notifications?limit=80&min_id=2340&types%5B%5D=mention HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/notifications?limit=80&min_id=2340&types%5B%5D=mention&types%5B%5D=status&types%5B%5D=reblog&types%5B%5D=follow&types%5B%5D=follow_request&types%5B%5D=favourite&types%5B%5D=poll&types%5B%5D=update HTTP/1.1" 200 2 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/timelines/home?limit=40&min_id=152252244464626184 HTTP/1.1" 200 6666 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/followed_tags?limit=80 HTTP/1.1" 404 3494 "-" "Ivory/12001 CFNetwork/1406.0.2 Darwin/22.4.0"


# mona beta 5 (38720)
## at launch
"GET /api/v1/accounts/verify_credentials HTTP/1.1" 200 1384 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v1/accounts/147050034222768154 HTTP/1.1" 200 1101 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v2/instance HTTP/1.1" 200 2147 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v1/timelines/home?limit=100 HTTP/1.1" 200 157524 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v1/conversations?limit=40 HTTP/1.1" 200 2 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v1/notifications?types%5B%5D=mention&types%5B%5D=reblog&types%5B%5D=favourite&types%5B%5D=follow&types%5B%5D=follow_request&types%5B%5D=poll&types%5B%5D=status&types%5B%5D=update&limit=40 HTTP/1.1" 200 29235 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"

## on refresh of home: 
"GET /api/v1/timelines/home?limit=100 HTTP/1.1" 200 152617 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"

## periodically: 
"GET /api/v1/conversations?limit=40 HTTP/1.1" 200 2 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"
"GET /api/v1/notifications?types%5B%5D=mention&types%5B%5D=reblog&types%5B%5D=favourite&types%5B%5D=follow&types%5B%5D=follow_request&types%5B%5D=poll&types%5B%5D=status&types%5B%5D=update&limit=40 HTTP/1.1" 200 29235 "-" "Mastodon/224 CFNetwork/1206 Darwin/20.1.0"



# mammoth 1.0.1 (113)
## on launch
"GET /media/profile_images/2023/2/11/Ir0pezv-TILY2a6xngc61Sp00fY.png HTTP/1.1" 200 0 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/lists HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/preferences HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/accounts/verify_credentials HTTP/1.1" 200 1384 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /media/profile_images/2023/2/11/Ir0pezv-TILY2a6xngc61Sp00fY.png HTTP/1.1" 200 0 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"DELETE /api/v1/push/subscription HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"DELETE /api/v1/push/subscription HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/lists HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"POST /api/v1/push/subscription HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/timelines/home HTTP/1.1" 200 77571 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"POST /api/v1/push/subscription HTTP/1.1" 404 3494 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"
"GET /api/v1/custom_emojis HTTP/1.1" 200 2 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"

## then every few seconds: 
"GET /api/v1/timelines/home?min_id=152252244464626184&limit=40 HTTP/1.1" 200 2 "-" "Mammoth/113 CFNetwork/1406.0.2 Darwin/22.4.0"

**Comment by dvydra** [comment]
Ok I'm watching Ivory (1.1) try and catch up on a couple of days worth of timeline right now. 

It's hitting the timelines/home api in groups of 6 requests then a 3 second pause in between groups, with a slowly incrementing min_id. 

As it does this, the unread post count pill in Ivory is increasing. It does 6 timeslines/home requests in a row, and the unread count increases by exactly 21 each time.

Hope this helps. This is easily reproducible so let me know if you'd like me to test anything else. 


Here is one set of 6 requests
```
127.0.0.1 - - [25/Feb/2023:12:02:27 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149197045286326521 HTTP/1.1" 200 171064 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
127.0.0.1 - - [25/Feb/2023:12:02:27 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149197410729991368 HTTP/1.1" 200 169395 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
127.0.0.1 - - [25/Feb/2023:12:02:28 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149197509986101152 HTTP/1.1" 200 169160 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
127.0.0.1 - - [25/Feb/2023:12:02:29 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149200391884813441 HTTP/1.1" 200 169460 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
127.0.0.1 - - [25/Feb/2023:12:02:29 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149200810613623184 HTTP/1.1" 200 169593 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
127.0.0.1 - - [25/Feb/2023:12:02:30 +0000] "GET /api/v1/timelines/home?limit=40&min_id=149201577369328768 HTTP/1.1" 200 169511 "-" "Ivory/11000 CFNetwork/1404.0.5 Darwin/22.3.0"
```

**Comment by dvydra** [comment]
SImilar story overnight. I opened Ivory on my iPhone and left it to run without interaction.
It created these requests. There was a significant amount of duplication, between 1 and 6 copies of each post. I can't see a pattern to why some things are duplicated more than others. 

[ivory.log](https://github.com/jointakahe/takahe/files/10832420/ivory.log)

**Comment by dcwatson** [comment]
I'm hoping #534 will fix this. Feel free to give it a shot and let me know, I'm running it on my instance for the next couple days to test.

**Comment by andrewgodwin** [comment]
I have also deployed #534 on takahe.social to test it there!

**Comment by dvydra** [comment]
I am running this on my instance and am unfortunately still seeing significant duplication. (And reversed ordering)

My account is following about 100 people.

**Comment by dcwatson** [comment]
Are you running the latest main as of today? I doubt it's in the latest docker image yet. Also, have you tried resetting your Ivory account cache? It's certainly still possible things are broken, I just haven't come across it yet (I have significantly fewer follows).

**Comment by dvydra** [comment]
I pulled main from GitHub and built the docker image from it last night. I did clear the ivory cache. I see you’ve pushed more related changes, I’ll try it again todaySent from my iPhoneOn 11 Mar 2023, at 09:24, Dan Watson ***@***.***> wrote:﻿
Are you running the latest main as of today? I doubt it's in the latest docker image yet. Also, have you tried resetting your Ivory account cache? It's certainly still possible things are broken, I just haven't come across it yet (I have significantly fewer follows).
